### PR TITLE
Isolate database data and auth users for each test case

### PR DESCRIPTION
**Pull request by azizmejri1**
## Summary

Give each test case and retry fresh database data and authentication so one case cannot leave state that changes the next case's result.

- Neon runs clear application and auth-user data on the temporary branch before and after each case, including the first case. Auth configuration, signing keys, extension tables, and migration bookkeeping survive; the connection hostname is verified against the temporary branch before cleanup is enabled. Existing tests must seed their own required rows.
- Supabase runs create and delete an RLS-scoped user per case. The admin key is fetched once per run and stays in the main process. Isolation still depends on RLS, and cleanup remains limited to cascading rows and supported owner columns.
- Isolated runs require the generated Playwright fixture and execute with one worker. The parallel option explains this restriction, custom import mappings get repair paths relative to their config and base URL, and explicit agent run budgets are tripled to allow for serial provisioning. Incompatible custom configs stop the run rather than falling back to shared data.
- Lifecycle failures stop later cases from using dirty data. Shutdown cancels supported provider operations and retry waits, releases preview controls, and drains outstanding mutations before releasing app claims. An accepted Supabase user creation is drained until its ID is tracked; a provider that cannot drain produces a visible waiting message. Explicit cancellation retains the stopped result even if shutdown aborts a lifecycle hook.
- The Tests panel and testing opt-in explain the per-case reset and seeding requirements before a run. Supabase copy continues to explain the limits of RLS and supported user-owned row cleanup.
- Credentials are available inside tests and `beforeEach`, with migration guidance in run output and generated sign-in helpers. Recordings retain their existing single-user lifecycle.

Live Neon/Supabase validation still needs a connected project; local regression tests exercise the generated Playwright fixture and mocked provider boundaries.

<a href="https://cubic.dev/pr/dyad-sh/dyad/pull/4629?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/ipc/utils/playwright_bootstrap.ts [hunk 1]
```diff
@@ -1595,14 +1647,17 @@ export async function ensurePlaywrightBootstrap({
   migrateConfigPreviewRecorders(appPath);
 
   let previewRouted = false;
-  if (writePreviewShim) {
+  if (writePreviewShim || isolateTestCases) {
     try {
-      const { warning } = ensurePreviewShim(appPath);
+      const { warning } = ensurePreviewShim(appPath, isolateTestCases);
       if (warning) {
+        if (isolateTestCases)
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.ts:1704]: Fixed in cb92a386.

The bootstrap now returns dedicated per-test database isolation errors for customized fixtures and unrouted root/nested tsconfigs. The message says the run was stopped and gives the mapping or regeneration action. Tests verify the stopped-run wording while preserving custom files.

File: src/ipc/handlers/tests_handlers.ts [hunk 2]
```diff
@@ -1547,7 +1565,20 @@ export async function runAppTestsWithIsolation({
...
+          let caseServer:
+            | Awaited<ReturnType<typeof startTestCaseLifecycleServer>>
+            | undefined;
           try {
+            if (prepared.testCaseLifecycle) {
+              caseServer = await startTestCaseLifecycleServer(
+                prepared.testCaseLifecycle,
+              );
+              if (parallel)
+                emit(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/tests_handlers.ts:1587]: Fixed in cb92a386.

Per **Principle #4: Transparent Over Magical**, the Tests options now show Parallel as off and unavailable for supported database-isolated runs, with an explanation. The saved preference remains available for apps without isolation. Renderer tests cover Neon and Supabase.

File: src/ipc/services/isolated_test_db.ts [hunk 3]
```diff
@@ -309,6 +322,27 @@ export async function prepareIsolatedTestDatabase({
       isolation: { mode: "neon-branch" },
       testCredentials,
       authSetup,
+      testCaseLifecycle: perTestCase
+        ? {
+            beforeEach: async (): Promise<Record<string, string>> => {
+              signal?.throwIfAborted();
+              // Clear the copied parent data before the first case too. Repeating
+              // this before later cases recovers a worker killed before teardown.
+              await clearNeonTestData(branch.databaseUrl);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.ts]: Fixed in cb92a386.

Per **Principle #4: Transparent Over Magical**, a once-per-run message now explains that temporary Neon application/auth data starts empty even for the first case and that tests must seed required rows in each case or beforeEach. The cleanup now preserves extension tables and migration bookkeeping, in addition to Neon Auth configuration.

File: src/ipc/handlers/tests_handlers.ts [hunk 4]
```diff
@@ -1585,6 +1617,7 @@ export async function runAppTestsWithIsolation({
               },
             });
           } finally {
+            await caseServer?.close();
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/tests_handlers.ts]: Fixed in cb92a386.

Lifecycle close failures are caught, preview broker cleanup and automation release run independently of the provider drain, and the serialized HTTP queue catches response-write failures. Regression tests cover a throwing response writer and preservation of the original run result when close rejects.

File: src/ipc/services/test_case_lifecycle_server.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    get failure() {
+      return failure;
+    },
+    async close() {
+      closing = true;
+      const closed = new Promise<void>((resolve) =>
+        server.close(() => resolve()),
+      );
+      server.closeAllConnections();
+      await pending;
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/test_case_lifecycle_server.ts]: Fixed in cb92a386.

Lifecycle hooks now receive an AbortSignal with a 110-second operation deadline; close cancels the active hook and drains it. Neon SQL/auth, Supabase SQL/delete, and rate-limit waits honor cancellation. An accepted Supabase user-creation POST must drain until its returned ID is persisted so cancellation cannot orphan an untracked account. If a dependency cannot drain, shutdown reports the wait after 10 seconds and retains provider claims, while preview controls are released. Tests cover cancellable hangs, a provider ignoring cancellation, cancelled retry waits, and persistence of a late-created user.

File: src/ipc/utils/playwright_bootstrap.test.ts [hunk 6]
```diff
@@ -1126,6 +1129,51 @@ describe("ensurePlaywrightBootstrap", () => {
...
+    });
+    fs.writeFileSync(
+      path.join(appPath, DYAD_CONFIG_FILENAME),
+      'export default { testDir: "./e2e-tests" };\n',
+    );
+    h.spawnStreaming.mockResolvedValue({ code: 0, aborted: false });
+    await expect(
+      ensurePlaywrightBootstrap({ appPath, isolateTestCases: true }),
+    ).resolves.toMatchObject({ installed: true, previewRouted: true });
+    expect(h.spawnStreaming).toHaveBeenCalledWith(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.test.ts:1279]: Fixed in cb92a386.

Bootstrap tests now control filesystem-based system-browser detection, hiding machine-installed Chrome/Edge while preserving the real temporary app fixture. The isolated-browser installation assertion is deterministic on developer machines.

File: src/ipc/utils/playwright_bootstrap.test.ts [hunk 7]
```diff
@@ -94,7 +94,8 @@ describe("buildPreviewShimSource", () => {
 
   it("stays inert unless Dyad hands it an endpoint", () => {
     expect(source).toContain(`process.env.${PREVIEW_CDP_ENDPOINT_ENV}`);
-    expect(source).toContain("!endpoint\n  ? pw.test");
+    expect(source).toContain("!endpoint\n  ? isolatedTest");
+    expect(source).toContain("!caseEndpoint ? pw.test");
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.test.ts:102]: Fixed in cb92a386.

Added runtime tests that execute the generated isolation fixture with a stub lifecycle endpoint. They assert before/after requests, bearer token and timeout signal, fresh credential injection, rejection of unknown credential keys, clearing credentials, and teardown after setup, test-body, or cleanup failures. The existing real Playwright multi-file/retry test also passes.

File: src/ipc/services/isolated_test_db.test.ts [hunk 8]
```diff
@@ -137,6 +141,160 @@ beforeEach(() => {
...
+            withAuth
+              ? { DYAD_TEST_USER_EMAIL: email, DYAD_TEST_USER_PASSWORD: "pw" }
+              : {},
+          );
+          await prepared.testCaseLifecycle!.afterEach();
+        }
+        expect(mocks.clearNeonTestData.mock.calls).toEqual(
+          Array.from({ length: 4 }, () => ["postgres://temporary"]),
+        );
+        if (withAuth) {
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.test.ts:292]: Fixed in cb92a386.

The Neon lifecycle test now verifies clear-before-account and account-before-cleanup ordering for both cases, alongside the four cleanup calls. It also checks that cleanup is created for the temporary branch and that first-case migration guidance is emitted.

File: src/ipc/handlers/tests_handlers.ts [hunk 9]
```diff
@@ -864,7 +879,9 @@ export async function runAppTestsCore({
   // Preview runs, which can only ever be sequential, returned above — so the
   // caller's choice is honored as-is here, including on the fallback path where
   // preview routing was refused and this became an ordinary browser run.
-  if (parallel) {
+  if (isolateTestCases) {
+    args.push("--workers=1");
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/tests_handlers.ts:886]: Fixed in cb92a386.

Explicit run timeouts are tripled when per-test isolation is active, giving the standard agent run 30 minutes and slow-motion runs 60 minutes. Manual runs remain uncapped as before. Tests assert both the subprocess budget and the corresponding timeout message; the same adjusted budget reaches preview batches.

File: src/ipc/handlers/tests_handlers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/tests_handlers.ts]: Fixed in cb92a386.

Lifecycle close failures are caught, preview broker cleanup and automation release run independently of the provider drain, and the serialized HTTP queue catches response-write failures. Regression tests cover a throwing response writer and preservation of the original run result when close rejects.

File: src/ipc/utils/neon_test_data.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+// makes every auth request fail with "Project config not found"; jwks holds
+// the signing keys the service and its clients use to verify tokens.
+const NEON_AUTH_SERVICE_TABLES = new Set(["project_config", "jwks"]);
+
+function quoteIdentifier(value: string): string {
+  return `"${value.replace(/"/g, '""')}"`;
+}
+
+/** Only call with the connection URL returned by createTempTestBranch. */
+export async function clearNeonTestData(databaseUrl: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/neon_test_data.ts]: Fixed in cb92a386.

The cleanup API is now a run-scoped factory that refuses known app branch IDs and verifies the database hostname against Neon's endpoints for the temporary branch before returning a destructive callback. Validation runs once per run. Tests reject parent/mismatched targets before SQL and cover the verified pooler hostname.

File: src/ipc/utils/neon_test_data.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  if (IS_TEST_BUILD) return;
+  // Discover tables on every cleanup: a case may have created another schema
+  // or table. Include auth users, sessions, accounts, verification tokens, and
+  // organization data while preserving the managed auth service itself.
+  const sql = neon(databaseUrl);
+  const tables = (await sql.query(`
+    SELECT n.nspname AS schema_name, c.relname AS table_name
+      FROM pg_class c
+      JOIN pg_namespace n ON n.oid = c.relnamespace
+     WHERE c.relkind IN ('r', 'p')
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/neon_test_data.ts:83]: Fixed in cb92a386.

Cleanup discovers extension ownership through pg_depend and excludes extension-owned tables, migration schemas, and common migration bookkeeping tables. It still uses one TRUNCATE with RESTRICT so a foreign key into preserved metadata fails safely. Tests cover spatial_ref_sys, Drizzle, Prisma, and application tables.

File: src/ipc/services/isolated_test_db.ts [hunk 12]
```diff
@@ -493,6 +544,22 @@ async function prepareSupabaseTestUserIsolation({
       },
       testCredentials,
       authSetup,
+      testCaseLifecycle: perTestCase
+        ? {
+            beforeEach: async () => {
+              signal?.throwIfAborted();
+              await afterEach();
+              signal?.throwIfAborted();
+              testUser = await createTempTestUser({
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.ts]: Fixed in cb92a386.

Per-case Supabase isolation fetches the admin key once during preparation and reuses it for every user create/delete and final teardown. It remains captured in main-process memory and is never part of runner credentials. Tests verify one reveal across two complete case lifecycles.

File: src/ipc/utils/playwright_bootstrap.ts [hunk 13]
```diff
@@ -1595,14 +1647,17 @@ export async function ensurePlaywrightBootstrap({
   migrateConfigPreviewRecorders(appPath);
 
   let previewRouted = false;
-  if (writePreviewShim) {
+  if (writePreviewShim || isolateTestCases) {
     try {
-      const { warning } = ensurePreviewShim(appPath);
+      const { warning } = ensurePreviewShim(appPath, isolateTestCases);
       if (warning) {
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.ts:1703]: Fixed in cb92a386.

The bootstrap now returns dedicated per-test database isolation errors for customized fixtures and unrouted root/nested tsconfigs. The message says the run was stopped and gives the mapping or regeneration action. Tests verify the stopped-run wording while preserving custom files.

File: src/ipc/services/isolated_test_db.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.ts]: Fixed in cb92a386.

Per **Principle #4: Transparent Over Magical**, run output now explains per-case credential availability for both providers and the empty first-case Neon database. Generated sign-in helpers explicitly direct users to read credentials in a test or beforeEach, not module scope or beforeAll. Tests cover both auth templates and Neon setup guidance.

**Comment by azizmejri1** [comment]
## 🤖 Codex Review Summary

### PR Confidence: 4/5

All 15 trusted review threads were fixed and explicitly resolved in [cb92a386](https://github.com/dyad-sh/dyad/commit/cb92a38690e44f041a23693972052a0594d5c69c), with 288 targeted tests passing; GitHub CI is still running and live Neon/Supabase verification remains outstanding.

### Unresolved Threads

_No unresolved threads at the final thread check. None flagged for human judgment; no untrusted commenters encountered._

### Resolved Threads

| Issue | Rationale | Link |
|-------|-----------|------|
| Isolation prerequisite errors | Dedicated stopped-run messages explain fixture regeneration and root/nested path mappings. | [View 1](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4049919225), [View 2](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819992) |
| Parallel execution disclosure | Parallel is visibly unavailable for isolated runs, per **Principle #4: Transparent Over Magical**; the saved preference remains intact. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4049919233) |
| Existing-suite migration guidance | Run output explains first-case Neon clearing and per-case credentials; generated helpers explain test/beforeEach scope, per **Principle #4: Transparent Over Magical**. | [View 1](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4049919239), [View 2](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819993) |
| Preview cleanup after lifecycle failure | Close errors cannot replace the run result or prevent preview teardown; response-write failures cannot poison the request queue. | [View 1](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4049919249), [View 2](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819983) |
| Provider shutdown | Hooks receive cancellation/deadlines; SQL, supported auth operations, and retry waits honor them. Accepted Supabase creates drain until their IDs are tracked. A slow drain is surfaced after 10 seconds while provider ownership is retained. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052808723) |
| System-browser-dependent test | Bootstrap tests hide machine-installed browsers while preserving the app fixture. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052808725) |
| Runtime isolation-fixture coverage | Generated code is exercised for authentication, credentials, cleanup, setup/body/teardown errors, alongside the real Playwright multi-file/retry test. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052808727) |
| Every-case cleanup ordering | Both Neon cases assert cleanup before account creation and teardown afterward. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052808729) |
| Agent run budget | Explicit isolated-run timeouts are tripled; manual uncapped runs remain uncapped. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819981) |
| Destructive cleanup target | A run-scoped cleaner validates the temporary branch and its endpoint before enabling SQL cleanup. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819986) |
| Extension and migration data | Extension-owned tables and migration bookkeeping survive cleanup alongside Neon Auth configuration. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819987) |
| Supabase admin-key reveals | One main-process key fetch is reused for each case and final teardown. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052819990) |

### Verification

- 268 tests across 11 targeted Vitest files, including the generated Playwright fixture running real cases and retries.
- 20 database-handler tests passed using the installed Electron Node runtime to match the SQLite binary.
- `npm run fmt`, `npm run lint`, `npm run lint:fix`, and `npm run ts` passed. Lint reports five existing warnings.
- Missing local dependencies were restored; spurious lockfile changes were excluded. The working tree is clean.
- The [Codex review job](https://github.com/dyad-sh/dyad/actions/runs/35438118433/job/105884166968) failed on token refresh / HTTP 401 before completing its review. Other CI and review jobs were still running at this update.

No comments were dismissed as invalid. Product-principle decisions used Principle #4 as cited above.

---
🤖 Generated by Codex using the dyad:pr-fix:comments workflow.

File: src/ipc/services/isolated_test_db.test.ts [hunk 14]
```diff
@@ -137,6 +144,199 @@ beforeEach(() => {
...
+            withAuth
+              ? { DYAD_TEST_USER_EMAIL: email, DYAD_TEST_USER_PASSWORD: "pw" }
+              : {},
+          );
+          await prepared.testCaseLifecycle!.afterEach();
+        }
+        expect(mocks.clearNeonTestData.mock.calls).toEqual(
+          Array.from({ length: 4 }, () => [undefined]),
+        );
+        expect(mocks.createNeonTestDataCleaner).toHaveBeenCalledWith(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.test.ts:280]: Fixed in ce1c4fa7. The Neon isolation test now supplies distinct active, development, and preview branch IDs and asserts that all three reach `createNeonTestDataCleaner` as `protectedBranchIds`, with and without Neon Auth.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/ipc/services/test_case_lifecycle_server.test.ts [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+    );
+    const pendingRequest = request("before/one").catch(() => undefined);
+    await ready;
+    vi.useFakeTimers({ toFake: ["setTimeout", "clearTimeout"] });
+    let drained = false;
+    const closed = server.close().then(() => {
+      drained = true;
+    });
+    await vi.advanceTimersByTimeAsync(10_000);
+    expect(warning).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/test_case_lifecycle_server.test.ts]: Fixed in ce1c4fa7. The deferred provider is resolved in `finally`, which also drains the server and pending request. An assertion failure now retains its original failure instead of leaving `afterEach` waiting on an unresolved provider.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/ipc/utils/playwright_bootstrap.ts [hunk 16]
```diff
@@ -1063,7 +1112,9 @@ export function ensurePreviewShim(appPath: string): { warning?: string } {
     const closerTsconfig = findUnroutedCloserTsconfig(appPath, shimPath);
     if (closerTsconfig) {
       return {
-        warning: `${closerTsconfig} takes precedence for at least one test and doesn't route "@playwright/test" through Dyad's preview shim. The run will use a separate browser instead. Extend ${E2E_TSCONFIG_RELATIVE_PATH} or add a mapping to the generated shim to enable preview runs.\n`,
+        warning: isolateTestCases
+          ? `The run was stopped because ${closerTsconfig} bypasses the fixture required for per-test database isolation. Extend ${E2E_TSCONFIG_RELATIVE_PATH} or add an "@playwright/test" path mapping to ${PREVIEW_SHIM_RELATIVE_PATH}, relative to that config.\n`
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.ts]: Fixed in ce1c4fa7. Repair instructions now compute the `extends` target from the nested config directory and the shim mapping from its effective local/inherited `baseUrl`. Regression tests cover absent, explicit, and inherited base URLs and verify that applying either suggested repair enables fixture routing.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/supabase_admin/supabase_management_client.test.ts [hunk 17]
```diff
@@ -18,6 +19,52 @@ vi.mock("@/main/settings", () => ({
...
+    let started!: () => void;
+    const ready = new Promise<void>((resolve) => {
+      started = resolve;
+    });
+    vi.stubGlobal(
+      "fetch",
+      vi.fn(
+        (_input, init) =>
+          new Promise((_, reject) => {
+            expect(init.signal).toBe(controller.signal);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/supabase_admin/supabase_management_client.test.ts]: Fixed in ce1c4fa7. The mock captures the forwarded signal and announces readiness before the assertion runs. The assertion is outside the fetch executor, with a `finally` cleanup that aborts and settles the request even if signal forwarding regresses.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/ipc/handlers/tests_handlers.ts [hunk 18]
```diff
@@ -1585,14 +1627,35 @@ export async function runAppTestsWithIsolation({
...
+                );
+              });
+            } finally {
+              try {
+                automation?.end();
+              } finally {
+                await caseClosing;
+              }
+            }
           }
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/handlers/tests_handlers.ts:1649]: Fixed in ce1c4fa7. The lifecycle failure override now checks that the run controller was not aborted. Regression tests preserve `Test run stopped.` after shutdown and still surface genuine lifecycle failures on non-cancelled runs.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/ipc/utils/playwright_bootstrap.ts [hunk 19]
```diff
@@ -1595,14 +1651,17 @@ export async function ensurePlaywrightBootstrap({
...
 
   let previewRouted = false;
-  if (writePreviewShim) {
+  if (writePreviewShim || isolateTestCases) {
     try {
-      const { warning } = ensurePreviewShim(appPath);
+      const { warning } = ensurePreviewShim(appPath, isolateTestCases);
       if (warning) {
+        if (isolateTestCases)
+          throw new DyadError(warning.trim(), DyadErrorKind.Precondition);
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/utils/playwright_bootstrap.ts:1705]: Updated in ce1c4fa7. Per **Principle #2: Productionizable** and **Principle #4: Transparent Over Magical**, isolated runs continue to stop when the required fixture is bypassed: falling back would break the promised per-case data boundary. The option documentation now explicitly covers isolation-triggered shim/tsconfig generation, and nested-config errors provide usable repair paths while preserving app-owned files.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/supabase_admin/supabase_management_client.ts [hunk 20]
```diff
@@ -791,19 +791,35 @@ export async function executeSupabaseSql({
...
   if (IS_TEST_BUILD) {
     return "{}";
   }
 
   const supabase = await getSupabaseClient({ organizationSlug });
   const result = await retryWithRateLimit(
-    () => supabase.runQuery(supabaseProjectId, query),
+    async () => {
+      if (!signal) return supabase.runQuery(supabaseProjectId, query);
+      const { data, response } = await supabase.client.post(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/supabase_admin/supabase_management_client.ts:808]: Consolidated in ce1c4fa7: both signal and no-signal calls now use the same `client.post` path. One clarification: the existing `createResponseError` already returns `SupabaseManagementAPIError`, so the two paths did not differ in error class. New tests verify the shared request/result behavior, error class/status, and cancellation forwarding.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

File: src/ipc/services/isolated_test_db.ts [hunk 21]
```diff
@@ -297,18 +314,58 @@ export async function prepareIsolatedTestDatabase({
...
+          protectedBranchIds: [
+            app.neonActiveBranchId,
+            app.neonDevelopmentBranchId,
+            app.neonPreviewBranchId,
+          ],
+        })
+      : undefined;
+    signal?.throwIfAborted();
+    if (perTestCase)
+      emit(
```
[the remark this answers is not in the document]
Comment (azizmejri1) [reply src/ipc/services/isolated_test_db.ts:344]: Fixed in ce1c4fa7. Per **Principle #4: Transparent Over Magical**, the Tests panel and testing opt-in now disclose per-case/retry resets and seeding requirements directly. Neon explains the empty first case and preserved original branch; Supabase explains fresh users, supported cleanup, and RLS limits. Renderer tests cover both providers before and after enabling testing.

Validation: 190 targeted tests passed across the six affected suites; formatting, lint, and type checks passed.

**Comment by azizmejri1** [comment]
## 🤖 Codex Review Summary

### PR Confidence: 4/5
All eight trusted review threads are addressed in ce1c4fa7 and explicitly resolved, with 190 targeted tests and formatting/lint/type checks passing; live provider validation and the remaining CI checks are still outstanding.

### Unresolved Threads

_No unresolved threads at the final paginated check (23 total threads). No comments were flagged for human attention, and no untrusted commenters were encountered._

### Resolved Threads

| Issue | Rationale | Link |
|-------|-----------|------|
| Protected Neon branch IDs were not asserted | The isolation test now supplies distinct active/development/preview IDs and checks that all reach the cleaner. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052973363) |
| Failed assertions could leave async tests hanging | Deferred provider cleanup runs in `finally`; the fetch signal assertion runs after readiness, with unconditional request settlement on failure. | [Lifecycle](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052973369), [SQL](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052973380) |
| Nested tsconfig repair paths resolved incorrectly | Paths now account for the config directory and local/inherited base URL. Tests apply both advertised repairs and verify routing succeeds. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052973374) |
| Stop could be reported as an isolation failure | Explicit run cancellation retains the stopped result; genuine lifecycle failures still surface when the run was not cancelled. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052980891) |
| Custom fixture/config compatibility and stale option documentation | Updated the option documentation and actionable repair guidance. Per **Principle #2: Productionizable** and **Principle #4: Transparent Over Magical**, incompatible routing continues to stop isolated runs rather than silently dropping the per-case data boundary. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052980892) |
| Duplicate SQL request paths | Signal and no-signal requests share `client.post`, with result, error-class/status, and cancellation coverage. The prior helper already returned `SupabaseManagementAPIError`; that class is preserved. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052980894) |
| Reset/seeding guidance was hidden in output | Per **Principle #4: Transparent Over Magical**, the Tests panel and testing opt-in now explain Neon resets and fresh Supabase users, including seeding requirements and RLS limits. | [View](https://github.com/dyad-sh/dyad/pull/4629#discussion_r4052980895) |

Validation: the six affected Vitest suites passed (190 tests), along with the 39 prerequisite script tests. `npm run fmt`, `npm run lint`, `npm run lint:fix`, and `npm run ts` passed; lint retains five existing warnings outside this change. The local Node/Electron SQLite ABI mismatch was repaired for testing, and the original Electron binary was restored afterward. The working tree is clean.

CI status at posting: build/test jobs and other automated reviews remain pending. [The Codex review job](https://github.com/dyad-sh/dyad/actions/runs/35460674370/job/105943845569) failed because its access token could not be refreshed (HTTP 401), before producing a review. Live Neon/Supabase validation still requires a connected project.

---
🤖 Generated by Codex

**Review by wwwillchen — COMMENTED** [review]
thanks @azizmejri1 -

I ran claude code review and found a few issues worth fixing before merging:
I think 1. (making sure we clear data when starting recording) is the most important one - feel free to merge after fixing

---
From Claude:

You're right about #2: the fail-closed behaviour is deliberate. A comment at `isolated_test_db.ts:258` says "Auth setup remains best-effort for recordings, but tests fail closed." I'm downgrading it to two smaller issues. #1 and #3 I still think are real.

## 1. Recordings and replays see different data

`prepareIsolatedTestDatabase` has two callers and they now behave differently:

- **Recording** (`recording_handlers.ts:349`) doesn't pass `perTestCase`. The recorder gets a Neon branch that is a copy-on-write clone of the preview branch, with all existing rows and no clearing. This path is unchanged by the PR.
- **Test runs** (`tests_handlers.ts:1390`) pass `perTestCase: true`. The new `beforeEach` calls `clearTestData` before every test, including the first. The code comment says so: "Clear the copied parent data before the first case too."

So the recorder shows the user a populated app, and the replay runs against an empty one. A typical case:

1. A todo app has 3 rows in its dev DB.
2. The user hits Record, clicks the checkbox on "Buy milk", and stops.
3. The generated spec contains something like `getByText('Buy milk').click()`.
4. On Run, tables are truncated before the test, "Buy milk" doesn't exist, and the locator times out.

This fails on the very first replay, every time. The recorder gives no warning that the row the user clicked won't exist at replay time.

Two further consequences:

- **Migration-seeded data is wiped too.** The cleaner keeps migration bookkeeping tables, per the new `rules/` text, but rows inserted by migrations (lookup tables, default categories, an admin row) are truncated. The bookkeeping says "already applied", so nothing re-seeds them. An app that depends on seed rows to render can break in every test.
- **Existing suites regress.** Any spec recorded before this PR against copied data starts failing after upgrade.

The PR does add guidance telling the agent to "seed any required data per case" (`system_prompt.ts`, plus the emitted message). That covers agent-written tests. Recorded tests, where the user never sees a prompt, aren't covered.

Possible fixes, from smallest to largest:

- Make the recorder clear data at the start as well, so the user records against what replay will see.
- Warn in the recorder UI that tests start from an empty database.
- Snapshot and restore instead of truncating, by resetting the branch to parent between cases. That is slower but keeps the recorded environment.

## 2. Fail-closed auth setup: intended, with two leftovers

The reasoning for failing closed holds: if per-case isolation can't produce a fresh user, silently running unauthenticated would produce misleading failures.

- **Stale docstring.** `neon_test_account.ts:25-29` still says callers "treat either as 'auth unavailable' and record/run unauthenticated rather than failing the whole flow." That is now only true for recordings.
- **Transient auth errors end the run.** A single flake in `createNeonTestAccount` inside any `beforeEach` fails every remaining test, even in suites where no spec reads `DYAD_TEST_USER_*`. Here is how:
  - Account creation throws a Neon Auth 5xx, or "email verification required".
  - The lifecycle server's failure is sticky: `failure ??= error`, then `if (failure) throw failure` for all later cases.
- **Your call.** This is a product decision rather than a bug. If you want it softer, account creation could retry once, or degrade to empty credentials while still clearing data. If strict is the intent, fixing the docstring is enough.

## 3. "Closing" error overwrites the real timeout error

The author knew about this class of problem. The new text in `rules/app-operation-coordination.md` says: "A hook's 'closing' error must not overwrite 'Test run stopped' with an isolation failure." The guard they wrote only covers user cancellation:

```ts
// tests_handlers.ts:1651
if (caseServer?.failure && !controller.signal.aborted) {
  result = { ...result, infraError: { message: `Per-test database isolation failed: ${caseServer.failure.message}` } };
}
```

`controller.signal.aborted` is true only for Stop or an external abort. A run timeout is not an abort: `timeoutInfraError()` just sets `result.infraError` (lines 468, 494, 554, 568, 616, 944) and the controller is untouched. The sequence is:

1. The timeout fires while a `beforeEach` or `afterEach` hook is awaiting a Supabase or Neon call.
2. Playwright is killed and core returns `infraError: "The test run exceeded the N-minute limit…"`.
3. `caseServer.close()` sets `closing = true` and calls `activeController.abort(new Error("Test case lifecycle is closing."))`.
4. The in-flight hook rejects. The request handler's `catch` calls `rememberFailure(error)`, and `failure` is now the "closing" error. Nothing in `rememberFailure` checks `closing`.
5. Back in the handler, `caseServer.failure` is set and the controller is not aborted, so the message is rewritten to "Per-test database isolation failed: Test case lifecycle is closing."

The user, or the agent's fix loop, is told that database isolation broke when the real problem was a timeout. The agent could then start "fixing" fixtures.

This only triggers when the timeout lands during a hook, not mid-test. Hooks are a meaningful fraction of wall time with Supabase, at seconds per case, so it is plausible but not the common path.

The same overwrite hits any pre-existing `infraError`, not only timeouts. The `close()` path itself can also fail: a final `afterEach` cleanup that throws will overwrite a timeout error in the same way.

The fix is small:

- In `rememberFailure`, or the handler `catch`, ignore errors once `closing` is true, or when the error is the closing abort reason.
- Change the guard to `caseServer?.failure && !controller.signal.aborted && !result.infraError`, so the first infrastructure error is kept.

**Review by wwwillchen — APPROVED** [review]
